examples/graffiti.js

```diff
@@ -32,37 +38,85 @@ bot.on('chat', (username, message) => {
   if (username === bot.username) return
   switch (true) {
     case /^watch$/.test(message):
-      watchPaintingOrSign()
+      watchSign()
       break
     case /^write .+$/.test(message):
       // write message
       // ex: write I love diamonds
       updateSign(message)
       break
+    case /^place .+$/.test(message):
+      placeSign(message)
   }
 })
 
-function watchPaintingOrSign () {
-  const paintingBlock = bot.findBlock({
-    matching (block) {
-      return !!block.painting
-    }
-  })
+function watchSign () {
   const signBlock = bot.findBlock({
-    matching: ['painting', 'sign'].map(name => mcData.blocksByName[name].id)
+    matching: mcData.blocksArray.filter(b => b.name.includes('sign')).map(b => b.id)
   })
   if (signBlock) {
     bot.chat(`The sign says: ${signBlock.signText}`)
-  } else if (paintingBlock) {
-    bot.chat(`The painting is: ${paintingBlock.painting.name}`)
   } else {
     bot.chat('There are no signs or paintings near me')
   }
 }
 
+async function placeSign (message) {
+  const signItem = bot.inventory.items().find(i => i.name.includes('sign'))
+  if (!signItem) {
+    bot.chat('Give me a sign first!')
+    return
+  }
+  await bot.equip(signItem, 'hand')
+  // Look for solid blocs first
+  const freeBlocks = bot.findBlocks({
+    matching: (block) => {
+      return block.boundingBox === 'block'
+    },
+    maxDistance: 4,
+    count: 100
+  })
+  // Then filter for blocks that have an air gap above them
+  const freeBlock = bot.blockAt(freeBlocks.find((pos) => {
+    const block = bot.blockAt(pos.offset(0, 1, 0))
+    return block && block.boundingBox === 'empty' && block.name.includes('air')
+  }))
+  if (!freeBlock) {
+    bot.chat('No free block to place a sign on')
+    return
+  }
+
+  const signEditPromise = (async () => {
+    // Wait for the right event to fire that indicates that we have opened our sign gui
+    // If we update the sign text before this event fires the server might reject the sign text
+    for await (const event of on(bot._client, 'packet')) {
```

Comment:
this is not good

Reply:
Can you be more specific? What is not good about it?

Reply:
User code shouldn't be using ._client or listening to packets

Reply:
Can you be more specific about what is not good about it?

Reply:
Indeed

Reply:
I changed it in the last commit